rename scan to scan_with_keychain

448-OG opened the pull request:
The purpose of this PR is to make changes regarding issue https://github.com/bitcoindevkit/bdk/issues/1112

### Notes to the reviewers

This is a draft pull request 

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing


#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing (I run the tests using `cargo test --all-features`
* [x] I'm linking the issue being fixed by this PR

448-OG commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -165,20 +165,15 @@ pub trait ElectrumExt<A> {
     ///
     /// [`scan`]: ElectrumExt::scan
-    fn scan_without_keychain(
+    fn sync(
```
Here :
```
fn sync(
        &self,
        prev_tip: Option<CheckPoint>,
        misc_spks: impl IntoIterator<Item = ScriptBuf>,
        txids: impl IntoIterator<Item = Txid>,
        outpoints: impl IntoIterator<Item = OutPoint>,
        batch_size: usize,
    )
    ```
    
    The method arguments `prev_tip, txids, outpoints, batch_size` are not used. Should I remove them also?

448-OG replied:
@LLFourn @vladimirfomene

LLFourn replied:
WIth sync you definitely need `prev_tip` and `batch_size` but I don't think you need `stop_gap` since you are meant to exhaustively iterate through `misc_spks`.

448-OG commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -191,38 +186,26 @@ pub trait ElectrumExt<A> {
 
-        let txids = txids.into_iter().collect::<Vec<_>>();
-        let outpoints = outpoints.into_iter().collect::<Vec<_>>();
+        let txids = Vec::<Txid>::new();
```
`sync_with_keychain` no longer has these fields, Should they remain empty `Vec`s or is there more refactor needed @vladimirfomene @LLFourn

danielabrozzoni replied:
Vec works fine here

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -165,20 +165,15 @@ pub trait ElectrumExt<A> {
     /// The scan for each keychain stops after a gap of `stop_gap` script pubkeys with no associated
     /// transactions. `batch_size` specifies the max number of script pubkeys to request for in a
     /// single batch request.
```
You should also improve the documentation explaining when to use this method, as described in #1112

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -165,20 +165,15 @@ pub trait ElectrumExt<A> {
         &self,
-        prev_tip: Option<CheckPoint>,
         keychain_spks: BTreeMap<K, impl IntoIterator<Item = (u32, ScriptBuf)>>,
-        txids: impl IntoIterator<Item = Txid>,
-        outpoints: impl IntoIterator<Item = OutPoint>,
```
@LLFourn when you said:

> I would remove all the other arguments from scan_with_keychain other than the keychain_spks in both esplora and electrum. This emphasizes the special case that it should be used for.

Did you mean to remove *all* the other arguments, or just txids/outpoints? Personally I think it makes sense to leave `prev_tip`, `stop_gap`, `batch_size`, but I might be missing something

LLFourn replied:
Yes sorry @448-OG I meant what @danielabrozzoni. `scan_with_keychain` should only be for script pubkeys but needs `stop_gap` and `batch_size` and `prev_tip`.

danielabrozzoni commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -231,35 +214,20 @@ impl ElectrumExt<ConfirmationHeightAnchor> for Client {
-                }
+                        keychain_spks,
+                        usize::default(),
+                        usize::default(),
```
I'm pretty sure we should keep `stop_gap` and `batch_size` as arguments. Using `usize::default()` instead produces the wrong result.

danielabrozzoni reviewed (changes requested):
Left a few comments. Tests are failing, check the CI logs :)

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -191,38 +186,26 @@ pub trait ElectrumExt<A> {
-            usize::MAX,
-            batch_size,
-        )
+        self.sync_with_keychain([((), spk_iter)].into())
```
This default implementation no longer makes sense The impl of sync cannot call `sync_with_keychain`.. There should be an internal function that is what `scan` used to be that both these functions call to achieve their functionality.

vladimirfomene replied:
@448-OG can you look into this?

LLFourn commented on crates/esplora/src/blocking_ext.rs:
```diff
@@ -46,7 +46,7 @@ pub trait EsploraExt {
     /// parallel.
     #[allow(clippy::result_large_err)]
-    fn update_tx_graph<K: Ord + Clone>(
+    fn sync_with_keychain<K: Ord + Clone>(
```
I think this should be called `scan_with_keychain`

448-OG replied:
This has been resolved

448-OG committed:
BREAKING CHANGE: In `electrum` crate the method `scan` of trait `ElectrumExt` no longer takes `txids` and `outpoints` as arguments

Fix: `example_electrum` and `wallet_electrum` in `example-crates` no longer take values for these arguments that implement the `scan` method of trait `ElectrumExt`

448-OG committed:
BREAKING CHANGE: In `electrum` crate rename the `ElectrumExt` trait method `scan` to `scan_with_keychain`

BREAKING CHANGE: In `electrum` crate rename the `ElectrumExt` trait method `scan_without_keychain` to `sync`

Fix: Modify the `example_electrum` and `wallet_electrum` examples to use new methods from `ElectrumExt` trait

448-OG committed:
Chore: Run cargo fmt

Chore: Implement recommendations from cargo clippy

448-OG commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -158,14 +158,12 @@ pub trait ElectrumExt<A> {
     /// transactions. `batch_size` specifies the max number of script pubkeys to request for in a
     /// single batch request.
-    fn scan<K: Ord + Clone>(
+    /// ##### NOTE: Scanning with keychain is very inefficient and should be used only when restoring
```
Is this wording okay if not how would I improve it

danielabrozzoni replied:
I would drop the "##### NOTE" and go with "Note that":

```
/// Note that scanning with keychain is very inefficient and should be used only when restoring
/// from seed words.
```

448-OG commented:
I have made some changes, a review is most welcome.

danielabrozzoni commented:
You have a few conflicts, rebase on master to fix them

448-OG commented:
> You have a few conflicts, rebase on master to fix them

Sorry about that. I have fixed the conflicts

448-OG committed:
BREAKING CHANGE: In `electrum` crate rename the `ElectrumExt` trait method `scan` to `scan_with_keychain`

BREAKING CHANGE: In `electrum` crate rename the `ElectrumExt` trait method `scan_without_keychain` to `sync`

Fix: Modify the `example_electrum` and `wallet_electrum` examples to use new methods from `ElectrumExt` trait

This commit has been rebased to use commits at hash `c20a4da9fc902a062eb217b39e37a32d9e64a148`

The requested updates have been applied using this commits and previous commits are mostly outdated

vladimirfomene commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -836,7 +836,7 @@ mod test {
         let target_amount = 250_000 + FEE_AMOUNT;
 
-        let result = LargestFirstCoinSelection::default()
+        let result = LargestFirstCoinSelection
```
I think these changes on coin selection slipped in because of the rebase. You have to remove them. There is more other files. Just cross check

448-OG replied:
These changes were as a result of running clippy. If you run clippy on master you will get recommendations to remove them. Since in rust tuple struct `LargestFirstCoinSelection` you dont need to call `::default()` to initialize it

notmandatory replied:
Generally it's better to handle fixing clippy changes in a separate commit or PR to make the review easier. Also a nit on commit messages, instead of `BREAKING CHANGE:` I'd use the notation `refactor!(<module name>): ...`. See: https://www.conventionalcommits.org/en/v1.0.0/

And as @LLFourn mentioned we may end up incorporating this rename into another PR but if you have time it wouldn't hurt to fix this one up in case we can use it.

vladimirfomene commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -139,59 +139,47 @@ pub trait ElectrumExt {
-        txids: impl IntoIterator<Item = Txid>,
-        outpoints: impl IntoIterator<Item = OutPoint>,
+        _txids: impl IntoIterator<Item = Txid>,
+        _outpoints: impl IntoIterator<Item = OutPoint>,
```
why don't we completely remove `_txids` and `_outpoints`? I don't see their utility

448-OG replied:
They are used in `bdk/example-crates/example_electrum/src/main.rs` line 249, do I remove the code in examples too to reflect this?

LLFourn replied:
`txids` and `outpoints` is meant to return you any relevant transactions to these items. They can't just be `_ignored`. A more significant refactoring is needed so that this method does not call `scan_with_keychain` but implements its own scanning logic.

vladimirfomene commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -139,59 +139,47 @@ pub trait ElectrumExt {
         prev_tip: Option<CheckPoint>,
         keychain_spks: BTreeMap<K, impl IntoIterator<Item = (u32, ScriptBuf)>>,
-        txids: impl IntoIterator<Item = Txid>,
-        outpoints: impl IntoIterator<Item = OutPoint>,
```
Now that we are removing `txids` and `outpoints` it doesn't make sense to be calling `populate_with_txids` and `populate_with_outpoints` here. Should separate methods be created to get these updates? @LLFourn @evanlinjin

LLFourn replied:
Yes the separate method is `scan`. `scan` is still meant to be doing this.

448-OG commented:
I will fix doc issues ASAP

danielabrozzoni reviewed (commented):
Make sure that all your commits are clean, and that don't introduce changes that you later undo in newer commits. For example, 2fa5fc606884b133ba86b5b166069c078b3c5b4b introduces a couple of unwanted "<<<<<<< HEAD" that you later remove in 2a9a2e0083bbede86e9deb1ab1ca7d7ee7e5099a - instead, squash the two commits together.

Regarding commit messages:
- We use conventional commits (https://www.conventionalcommits.org/en/v1.0.0/), which use the `!` in the commit title to signal breaking changes (https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with--to-draw-attention-to-breaking-change). Also, we don't put "In ... crate", we simply put "*name of the crate*:" at the start of the title. As an example, 7e91849f7c1896d2bdc4affc329eb50b3411828c's title should go from:
```
BREAKING CHANGE: In electrum crate the method scan of trait `Elec…
…trumExt` no longer takes `txids` and `outpoints` as arguments
```
to:
```
refactor(electrum)!: `ElectrumExt::scan` no longer takes `txids`, `outpoints`
```

Notice that the part before `:` is in lowercase characters, and the part after `:` is capitalized.
9655e315b41b14282847e29d660b76fc1fb84785's now looks like this:
```
Chore: Run cargo fmt
```
instead, it should be:
```
chore: Run cargo clippy
```
Since you're running clippy and not fmt :) Also, if the commit title is explanatory enough, you don't have to write a commit message.

If you've never rewritten commit messages, this can be of help: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

448-OG committed:
Fix: Replace previous docs `ElectrumExt::scan` with `ElectrumClient::scan_with_keychain`

    This solves the issue where CI tests for docs.rs error on rustdoc warning `--cfg docsrs -Dwarnings`

448-OG commented:
> Make sure that all your commits are clean, and that don't introduce changes that you later undo in newer commits. For example, [2fa5fc6](https://github.com/bitcoindevkit/bdk/commit/2fa5fc606884b133ba86b5b166069c078b3c5b4b) introduces a couple of unwanted "<<<<<<< HEAD" that you later remove in [2a9a2e0](https://github.com/bitcoindevkit/bdk/commit/2a9a2e0083bbede86e9deb1ab1ca7d7ee7e5099a) - instead, squash the two commits together.
> 
> Regarding commit messages:
> 
>     * We use conventional commits (https://www.conventionalcommits.org/en/v1.0.0/), which use the `!` in the commit title to signal breaking changes (https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with--to-draw-attention-to-breaking-change). Also, we don't put "In ... crate", we simply put "_name of the crate_:" at the start of the title. As an example, [7e91849](https://github.com/bitcoindevkit/bdk/commit/7e91849f7c1896d2bdc4affc329eb50b3411828c)'s title should go from:
> 
> 
> ```
> BREAKING CHANGE: In electrum crate the method scan of trait `Elec…
> …trumExt` no longer takes `txids` and `outpoints` as arguments
> ```
> 
> to:
> 
> ```
> refactor(electrum)!: `ElectrumExt::scan` no longer takes `txids`, `outpoints`
> ```
> 
> Notice that the part before `:` is in lowercase characters, and the part after `:` is capitalized. [9655e31](https://github.com/bitcoindevkit/bdk/commit/9655e315b41b14282847e29d660b76fc1fb84785)'s now looks like this:
> 
> ```
> Chore: Run cargo fmt
> ```
> 
> instead, it should be:
> 
> ```
> chore: Run cargo clippy
> ```
> 
> Since you're running clippy and not fmt :) Also, if the commit title is explanatory enough, you don't have to write a commit message.
> 
> If you've never rewritten commit messages, this can be of help: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

Thanks. I will look into commit messages

448-OG committed:
docs(electrum): Dropped usage of `NOTE` and replaced it with `Note that ` in the

danielabrozzoni commented:
Hey @448-OG, are you still working on this? Do you need help with anything?

448-OG commented:
> Hey @448-OG, are you still working on this? Do you need help with anything?

Hey @danielabrozzoni  I had finished doing changes, there were some comments @vladimirfomene made that I commented on. Unless there are new changes requested, my PR is ready for review

nondiremanuel commented:
After today's discussion in the Lib Team Call, a check on whether all the comments were addressed is needed on this PR. Also it will probably need to be rebased after the ci fix of #1182

LLFourn commented:
My opinion is that we probably want to close this PR in favor of one that solves all the problems in #1195 at once.

notmandatory commented:
@448-OG thanks for your work on this but the issue got fixed with #1235.

448-OG commented:
Thanks for the opportunity and reviews I got, I learnt quite a few things about the library.
